src/array_api_extra/_funcs.py

```diff
@@ -538,3 +548,314 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+            msg = (
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+            raise ValueError(msg)
+
+        if copy not in (True, False, None):
+            msg = f"copy must be True, False, or None; got {copy!r}"  # pyright: ignore[reportUnreachable]
+            raise ValueError(msg)
+
+        if copy is None:
+            writeable = is_writeable_array(x)
+            copy = not writeable
+        elif copy:
+            writeable = None
+        else:
+            writeable = is_writeable_array(x)
+
+        if copy:
+            if is_jax_array(x):
+                # Use JAX's at[]
+                func = getattr(x.at[idx], at_op)
+                return func(y), None
+            # Emulate at[] behaviour for non-JAX arrays
+            # with a copy followed by an update
+            if xp is None:
+                xp = array_namespace(x)
+            x = xp.asarray(x, copy=True)
+            if writeable is False:
+                # A copy of a read-only numpy array is writeable
+                # Note: this assumes that a copy of a writeable array is writeable
+                writeable = None
+
+        if writeable is None:
+            writeable = is_writeable_array(x)
+        if not writeable:
+            # sparse crashes here
+            msg = f"Can't update read-only array {x}"
+            raise ValueError(msg)
+
+        return None, x
+
+    def set(
+        self,
+        y: Array,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+    ) -> Array:  # numpydoc ignore=PR01,RT01
+        """Apply ``x[idx] = y`` and return the update array."""
+        res, x = self._update_common("set", y, copy=copy, xp=xp)
+        if res is not None:
+            return res
+        assert x is not None
+        x[self._idx] = y
+        return x
+
+    def _iop(
+        self,
+        at_op: Literal[
+            "set", "add", "subtract", "multiply", "divide", "power", "min", "max"
+        ],
+        elwise_op: Callable[[Array, Array], Array],
+        y: Array,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+    ) -> Array:  # numpydoc ignore=PR01,RT01
+        """
+        ``x[idx] += y`` or equivalent in-place operation on a subset of x.
+
+        which is the same as saying
+            x[idx] = x[idx] + y
+        Note that this is not the same as
+            operator.iadd(x[idx], y)
+        Consider for example when x is a numpy array and idx is a fancy index, which
+        triggers a deep copy on __getitem__.
+        """
+        res, x = self._update_common(at_op, y, copy=copy, xp=xp)
+        if res is not None:
+            return res
+        assert x is not None
+        x[self._idx] = elwise_op(x[self._idx], y)
```

Comment:
As currently implemented, the `_iop` path has different semantics for repeated indices between JAX and NumPy:
```python
>>> import numpy as np
>>> x = np.zeros(4)
>>> idx = np.array([1, 2, 2, 3, 3, 3])
>>> x[idx] = x[idx] + 1
>>> x
array([0., 1., 1., 1.])

>>> import jax.numpy as jnp
>>> x = jnp.zeros(4)
>>> idx = jnp.array([1, 2, 2, 3, 3, 3])
>>> x.at[idx].add(1)
Array([0., 1., 2., 3.], dtype=float32)
```
At the very least, the difference should be documented.

Reply:
it's already documented at lines 612:622